### DISCUSSION TITLE
Is it possible to create an automatic table with fluentmigrator?

### DISCUSSION THREAD

**Discussion by suatsuphi** [fluentmigrator/fluentmigrator#1805] [Q&A] [answered]
Hi,

We have a sample table like below. Is it possible to create the following tables with migration? Is there a command that all indexes and constraints will be automatic? Isn't it very unnecessary to define them one by one for initial? 

```
[MigrationVersion("2024-05-15 14:00:00", "Initial")]
internal class Initial : Migration
{
       if (!Schema.Table(nameof(Post)).Exists())
       {
              Create.Table(nameof(Post)).Automatic(Post);
       }
       if (!Schema.Table(nameof(Tag)).Exists())
       {
              Create.Table(nameof(Tag)).Automatic(Tag);
       }
       if (!Schema.Table(nameof(PostTag)).Exists())
       {
              Create.Table(nameof(PostTag)).Automatic(Post);
       }
}

```

```
using System.Collections.Generic;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;

namespace Domain
{
    public class Post
    {
        public int Id { get; set; }

        [StringLength(400)]
        [Required]
        public string Name { get; set; }
        public List<PostTag> PostTags { get; } 
        public List<Tag> Tags { get; } 
    }

    public class Tag
    {
        public int Id { get; set; }

        [StringLength(400)]
        [Required]
        public string Name { get; set; }
        public List<PostTag> PostTags { get; } 
        public List<Post> Posts { get; } 
    }

    internal class PostTagMap : IEntityTypeConfiguration<PostTag>
    {
        public void Configure(EntityTypeBuilder<PostTag> builder)
        {
            builder
                .HasOne(a => a.Tag)
                .WithMany(a => a.PostTags)
                .HasForeignKey(c => c.PostsId)
                .OnDelete(DeleteBehavior.ClientSetNull);

            builder
               .HasOne(a => a.Tag)
               .WithMany(a => a.PostTags)
               .HasForeignKey(c => c.TagsId)
               .OnDelete(DeleteBehavior.ClientSetNull);
        }
    }

    [Index(nameof(PostsId), Name = "IX_PostTag_PostsId")]
    [Index(nameof(TagsId), Name = "IX_PostTag_PostsId")]
    [Index(nameof(DisplayOrder), Name = "IX_PostTag_PostsId")]
    [Index(nameof(PostsId), nameof(TagsId), IsUnique = true, Name = "IX_PostTag_PostsId_TagsId")]
    public class PostTag
    {
        public int Id { get; set; } 
        public int DisplayOrder { get; set; }
        public int PostsId { get; set; }
        public int TagsId { get; set; }
        public Post Post { get; set; } = null!;
        public Tag Tag { get; set; } = null!;
    }
}

```

**Comment by PhenX** [comment, marked as answer]
While it seems like a good idea at first, it will feel like a bullet in the foot after.

Let's say your model classes evolve: you'll have to add new migrations to change the database structure, and that's OK.
But how will you handle applying theses migrations on a newly created database with the new model classes ? Fluent Migrator will have already created the latest DB structure AND will try to apply the other migrations : it will fail.

One better approach is to generate initial migrations once and for all

**Reply by jzabroski** [reply]
Agree; I added my own answer here https://github.com/fluentmigrator/fluentmigrator/discussions/1805#discussioncomment-14226820

**Comment by jzabroski** [comment]
I usually create a baseline CreateDatabase.sql script for test environments where database unit tests run, using Microsoft SMO library.

In this way, I do not really have to have automatic initialization in a first migration script. I just assume a baseline. In fact, I periodically archive migrations and generate a new baseline; this is a trick I borrowed from RoundhousE back when we used that approach at erecruit in 2012. I believe RedGate Ready To Roll does something similar, as does Flyway (Java language-based migrations framework).

In terms of scaffolding the migration itself from a type definition, Quentin Starin's EasyMigrator (linked in the Readme) already does this but I do not believe he has kept the TargetFrameworks up to date. https://www.nuget.org/packages/EasyMigrator.FluentMigrator; it was last updated in 2021 and targets FluentMigrator >= 1.6.2, which is a bit erroneous even if the APIs are largely unchanged.

In terms of API I would like? I would like Migrations to be able to assume an abstract pre-condition that describes the schema of the database prior to running migrations. This is essentially what Entity Framework does when they serialize the model state into a database table for automatic migrations, but a bit better in that there is a cleaner way to verify merge conflicts. I've not actually built this since its a lot of work and I don't know how many people would appreciate it. Most engineers hate databases and want to write C# code.

**Comment by suatsuphi** [comment]
two times code needs to be written, whereas all definitions are made in the class, it is necessary to do this once for migration.

example: 

```
[Table("Notification")]
[Index(nameof(AuthorId), Name = "IX_Notification_AuthorId")]
[Index(nameof(Published), Name = "IX_Notification_Published")]
public class Notification : BaseEntity
{
    public int AuthorId { get; set; }

    public DateTime Published { get; set; }

    [MaxLength]
    public string Message { get; set; }
}

```

and these

```
[MigrationVersion("2022-12-14 10:34:22", "HelloWorld: Initial")]
public class _20221214103422_Initial : Migration
{
    public override void Up()
    {
        // The table name is taken from Domain->Attribute->Table
        var tableName = "Notification";
        
        if (!Schema.Table(tableName).Exists())
        {
            Create.Table(tableName)
                .WithIdColumn() // Adds the Id property as the primary key.
                .WithColumn(nameof(Notification.AuthorId))
                    .AsInt32()
                    .NotNullable()
                    .Indexed("IX_Notification_AuthorId")
                .WithColumn(nameof(Notification.Published))
                    .AsDateTime2()
                    .NotNullable()
                    .Indexed("IX_Notification_Published")
                .WithColumn(nameof(Notification.Message))
                    .AsMaxString()
                    .NotNullable();
        }
    }

    public override void Down()
    {
        // Ignore this for now.
    }
}

```



The following code should be sufficient, there is relevant information in the class. 

```
 
if (!Schema.Table(Notification).Exists())
{
    Create.Table(Notification);
}

```


in software logic it is necessary not to repeat something :) 


thanks

**Comment by suatsuphi** [comment]
In the migration of the table and / or filed, it should be checked whether it is the old version or the new version!

LIKE: if (!Schema.Table(tableName).Exists())

EXAMPLE:

```
[MigrationVersion("2023-05-25 12:30:00", "New Column")]
internal class RenameTextMessageToMessage : Migration
{
    const string tableName = "Notification";

    public override void Up()
    {
        if (!Schema.Table(tableName).Exists())
          if (Schema.Table(tableName).Column("TextMessage").Exists())
          {
              Rename.Column("TextMessage")
                  .OnTable(tableName)
                  .To("Message");
          }
    }

    public override void Down()
    {
 
    }
}
```


If it is desired to be done without breaking the existing structure, something like `Create.Table(Notification).Auto()` can be done

like: 

`Create.Table(Notification).AutoTableStructureByClass()`
